pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java

```diff
@@ -288,38 +295,51 @@ private boolean registerNamespace(String namespace, boolean ensureOwned) throws
             LOG.debug("findBrokerServiceUrl: {} - read-only: {}", bundle, readOnly);
         }
 
-        CompletableFuture<Optional<LookupResult>> future = new CompletableFuture<>();
+        ConcurrentOpenHashMap<NamespaceBundle, CompletableFuture<Optional<LookupResult>>> targetMap;
+        if (authoritative) {
+            targetMap = findingBundlesAuth;
+        } else {
+            targetMap = findingBundlesNoAuth;
+        }
 
-        // First check if we or someone else already owns the bundle
-        ownershipCache.getOwnerAsync(bundle).thenAccept(nsData -> {
-            if (!nsData.isPresent()) {
-                // No one owns this bundle
+        return targetMap.computeIfAbsent(bundle, (k) -> {
+            CompletableFuture<Optional<LookupResult>> future = new CompletableFuture<>();
 
-                if (readOnly) {
-                    // Do not attempt to acquire ownership
-                    future.complete(Optional.empty());
-                } else {
-                    // Now, no one owns the namespace yet. Hence, we will try to dynamically assign it
-                    pulsar.getExecutor().execute(() -> {
-                        searchForCandidateBroker(bundle, future, authoritative);
-                    });
-                }
-            } else if (nsData.get().isDisabled()) {
-                future.completeExceptionally(
+            // First check if we or someone else already owns the bundle
+            ownershipCache.getOwnerAsync(bundle).thenAccept(nsData -> {
```

Comment:
My idea was to just convert `owershipCache` value from `OwnedBundle` to a `CompletableFuture<OwnedBundle>`. 

The concern here is that `findingBundlesAuth` and `ownershipCache` can be out if sync after the removal, since the cleanup is done in background.

Reply:
Thanks @merlimat. Sorry not very clear for the idea. Would you please help share more info? After the convert in `ownershipCache`, what is the process logic in `findBrokerServiceUrl` or `createLookupResult`?

It seems that in `ownershipCache`, `AsyncLoadingCache<String, OwnedBundle> ownedBundlesCache` and `ConcurrentHashMap<String, CompletableFuture<OwnedBundle>> ownedBundlesCache` behaviours the same, they both have trick to make only once handling for zk path and ownedBundle creating.